### PR TITLE
  Additions for ORA2 instructor override score feature.

### PR DESCRIPTION
**Pull request by dcadams**
```
ORA2 has been modified in a seperate PR to include functionality
allowing the instructor to override a student's ORA score.

This override score is stored in Submissions as a Score without
a related submission. API methods have been added to create and
get the override score.
```

"

### CODE REVIEW

**Comment by dcadams** [comment]
Although I believe this is complete, this is a place holder PR for now until I have the ORA2 bit ready.

**Comment by dcadams** [comment]
@stvstnfrd @jbau at your convenience.

**Comment by dcadams** [comment]
@kluo at your convenience.

File: submissions/models.py [hunk 1]
```diff
@@ -205,6 +205,25 @@ def create_reset_score(cls, student_item):
...
+            points_possible (string): The max points possible for this submission.
+            o
+
+        """
+
+        return cls.objects.create(
+            student_item=student_item_model,
+            submission=None,
+            points_earned=float(override_grade),
+            points_possible=float(points_possible),
```
Comment (stvstnfrd) [inline submissions/models.py:225]: What's the incoming data type of `points_possible` and `override_grade`?

File: submissions/models.py [hunk 2]
```diff
@@ -205,6 +205,25 @@ def create_reset_score(cls, student_item):
...
 
+    @classmethod
+    def create_override_score(cls, student_item_model, points_possible, override_grade):
+        """
+        Create an "override" score (a score with a null submission and reset False).
+
+        Args:
+            student_item_model (StudentItem): The student item model.
+            points_possible (string): The max points possible for this submission.
+            o
```
Comment (stvstnfrd) [inline submissions/models.py]: It looks like you forgot to finish documentation?

Comment (stvstnfrd) [inline submissions/models.py]: redacted

File: submissions/tests/test_models.py [hunk 3]
```diff
@@ -32,7 +32,6 @@ def test_latest(self):
         latest_score = ScoreSummary.objects.get(student_item=item).latest
         self.assertEqual(second_score, latest_score)
 
-
```
Comment (stvstnfrd) [inline submissions/tests/test_models.py:35]: Can we remove this unrelated whitespace change to minimize diff/drift.

File: submissions/tests/test_models.py [hunk 4]
```diff
@@ -164,3 +163,39 @@ def test_highest_score_hidden(self):
         highest = ScoreSummary.objects.get(student_item=item).highest
         self.assertEqual(highest.points_earned, 1)
         self.assertEqual(highest.points_possible, 2)
+
+    def test_override_highest_no_score(self):
+        item = StudentItem.objects.create(
+            student_id="score_test_student",
+            course_id="score_test_course",
+            item_id="i4x://mycourse/special_presentation"
```
Comment (stvstnfrd) [inline submissions/tests/test_models.py]: Add a trailing comma.

File: submissions/tests/test_models.py [hunk 5]
```diff
@@ -164,3 +163,39 @@ def test_highest_score_hidden(self):
         highest = ScoreSummary.objects.get(student_item=item).highest
         self.assertEqual(highest.points_earned, 1)
         self.assertEqual(highest.points_possible, 2)
+
+    def test_override_highest_no_score(self):
+        item = StudentItem.objects.create(
```
Comment (stvstnfrd) [inline submissions/tests/test_models.py:169]: We're creating this same item in multiple locations. Could we consolidate into either a helper method or a test case `setUp`?

File: submissions/tests/test_models.py [hunk 6]
```diff
@@ -164,3 +163,39 @@ def test_highest_score_hidden(self):
...
+        )
+        Score.objects.create(
+            student_item=item,
+            submission=submission,
+            points_earned=1,
+            points_possible=15,
+        )
+
+        Score.create_override_score(item, 15, 5)
+        highest = ScoreSummary.objects.get(student_item=item).highest
```
Comment (stvstnfrd) [inline submissions/tests/test_models.py:200]: How does the scoring logic work; does it always take the highest score?

File: submissions/models.py [hunk 7]
```diff
@@ -205,6 +205,25 @@ def create_reset_score(cls, student_item):
             reset=True,
         )
 
+    @classmethod
+    def create_override_score(cls, student_item_model, points_possible, override_grade):
```
Comment (stvstnfrd) [inline submissions/models.py]: Is there a reason we're passing the `points_possible` before passing in
`points_earned`? This may be personal bias, but I'm used to seeing
earned points before possible points; e.g., "5 out of 10" and "5/10".

Also, it looks like the API's existing `set_score` method takes
`points_earned` first, followed by `points_possible`.

For consistency, could we update the signature to:

``` python
def create_override_score(cls, student_item, points_earned, points_possible):
    pass
```

File: submissions/tests/test_override_score.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+        # Simulate a database error when creating the override score
+        create_mock.side_effect = DatabaseError("Test error")
+        with self.assertRaises(sub_api.SubmissionInternalError):
+            sub_api.score_override(
+                self.STUDENT_ITEM,
+                10,
+                7,
+            )
+
+    def test_override_doesnt_overwrite_subm_score(self):
```
Comment (stvstnfrd) [inline submissions/tests/test_override_score.py]: I'm assuming `subm_score` is short for `submission_score`? Could we be explicit and rename it as such?

**Comment by stvstnfrd** [comment]
@dcadams Could you link to the ORA2 PR so we can see how these fit together?

File: submissions/models.py (hunk 1 above)
Comment (stvstnfrd) [inline submissions/models.py:225]: Ah, I see above that they're originally strings; should we have some form of error handling?

File: submissions/tests/test_models.py [hunk 9]
```diff
@@ -164,3 +163,39 @@ def test_highest_score_hidden(self):
...
+        Score.create_override_score(item, 10, 9)
+        highest = ScoreSummary.objects.get(student_item=item).highest
+        self.assertEqual(highest.points_earned, 9)
+        self.assertEqual(highest.points_possible, 10)
+
+    def test_override_highest_w_score(self):
+        item = StudentItem.objects.create(
+            student_id="score_test_student",
+            course_id="score_test_course",
+            item_id="i4x://mycourse/special_presentation"
```
Comment (stvstnfrd) [inline submissions/tests/test_models.py]: Add a trailing comma.

File: submissions/tests/test_override_score.py [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+        sub_api.score_override(
+            self.STUDENT_ITEM,
+            10,
+            8,
+        )
+
+        self.assertEqual(sub_api.get_score(self.STUDENT_ITEM)['points_earned'], 8)
+        self.assertEqual(sub_api.get_score(self.STUDENT_ITEM)['points_possible'], 10)
+
+    def test_override_with_one_score(self):
```
Comment (stvstnfrd) [inline submissions/tests/test_override_score.py:42]: I'm unfamiliar w/ the scoring logic; is there a difference between a submission with a single score versus one with multiple scores? Should we test the "many" case?

File: submissions/tests/test_override_score.py [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+        )
+
+        self.assertEqual(sub_api.get_score(self.STUDENT_ITEM)['points_earned'], 5)
+        self.assertEqual(sub_api.get_score(self.STUDENT_ITEM)['points_possible'], 10)
+
+    @patch.object(Score.objects, 'create')
+    def test_database_error(self, create_mock):
+
+        # Simulate a database error when creating the override score
+        create_mock.side_effect = DatabaseError("Test error")
```
Comment (stvstnfrd) [inline submissions/tests/test_override_score.py]: Since we're using single quotes everywhere else, can we switch to them here for consistency?

File: submissions/tests/test_override_score.py [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+            10,
+            5,
+        )
+
+        self.assertEqual(sub_api.get_score(self.STUDENT_ITEM)['points_earned'], 5)
+        self.assertEqual(sub_api.get_score(self.STUDENT_ITEM)['points_possible'], 10)
+
+    @patch.object(Score.objects, 'create')
+    def test_database_error(self, create_mock):
+
```
Comment (stvstnfrd) [inline submissions/tests/test_override_score.py]: Elsewhere in this file, there's no newline between the `def` line and the first inline comment; can we remove this empty line for consistency?

File: submissions/tests/test_models.py [hunk 13]
```diff
@@ -164,3 +163,39 @@ def test_highest_score_hidden(self):
...
+
+        # Override score with no score
+        Score.create_override_score(item, 10, 9)
+        highest = ScoreSummary.objects.get(student_item=item).highest
+        self.assertEqual(highest.points_earned, 9)
+        self.assertEqual(highest.points_possible, 10)
+
+    def test_override_highest_w_score(self):
+        item = StudentItem.objects.create(
+            student_id="score_test_student",
```
Comment (stvstnfrd) [inline submissions/tests/test_models.py:183]: Since we're using single quotes everywhere else, can we switch to them here for consistency?

File: submissions/api.py [hunk 14]
```diff
@@ -720,6 +720,68 @@ def set_score(submission_uuid, points_earned, points_possible):
...
+            u" item {item_id} in course {course_id} for student {student_id}"
+        ).format(
+            item_id=student_item_dict['item_id'],
+            course_id=student_item_dict['course_id'],
+            student_id=student_item_dict['student_id'],
+        )
+        logger.exception(msg)
+        raise SubmissionInternalError(msg)
+    else:
+        msg = u"Score overridden for item {item_id} in course {course_id} for student {student_id}".format(
```
Comment (stvstnfrd) [inline submissions/api.py]: While I normally rant "single-quotes FTW!", I just want to emphasize that I _do_ like double-quotes here. Following shell convention, double quotes are great for representing interpolated strings.

Carry on!

**Comment by stvstnfrd** [comment]
A few minor points, and a question or two, but otherwise this is looking good.

Is the companion Stanford-Online/edx-ora2#7 ready for review?

File: submissions/tests/test_models.py (hunk 5 above)
Comment (dcadams) [inline submissions/tests/test_models.py:169]: That makes sense in isolation but existing tests have the same lines and so I should really create a setup and modify existing tests; which you don't like. Please advise.

File: submissions/tests/test_models.py (hunk 6 above)
Comment (dcadams) [inline submissions/tests/test_models.py:200]: Scoring always takes the last score; whether or not the old implementation (coursewarestudentmodule) or the new implementation (Submissions repo). However, submissions has a built in submissions_scoresummary table that has foreign keys to highest and latest ids. Note in submissions, scores are immutable whereas csm only keeps the latest and previous are rolled off into coursestudentmodulehistory. That said, the summary table is being populated but not used anywhere just yet.

File: submissions/tests/test_override_score.py (hunk 10 above)
Comment (dcadams) [inline submissions/tests/test_override_score.py:42]: I saw that test too for reset. Since we're always getting the latest it doesn't seem necessary for overrides.

File: submissions/tests/test_models.py (hunk 13 above)
Comment (dcadams) [inline submissions/tests/test_models.py:183]: It wouldn't be consistent with what already exists in the same file.

File: submissions/tests/test_models.py (hunk 5 above)
Comment (stvstnfrd) [inline submissions/tests/test_models.py:169]: Gotcha. We can leave this as-is for now.

I've already created a backlog task to refactor this for upstream at a later date.

**Comment by kluo** [comment]
LGTM, but the commit message still has an incorrect link to PR 7 of this repo, instead of the edx-ora2 one.
